<a href="https://colab.research.google.com/github/Jayveersinh-Raj/Tensors_tensorflow/blob/main/0_tensorflow.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Fundamental conepts of tensor using tensorflow

Index: <br>
1) Intro. to tensors <br>
2) Getting info. from tensors <br>
3) Manipulation of tensors <br>
4) Tensors and Numpy <br>
5) Using @tf.function to fasten normal python functions<br>
6) GPUs and TPUs with tensorflow


#1) Intro. to tensors

In [2]:
# Tensorflow import
import tensorflow as tf #tf as alias, but can be aribtary, however tf is the norm

#Just checking the version
print(tf.__version__)

2.8.2


In [ ]:
#To create tensor using tf.constant()

#There are multiple ways to create tensors however, but further, we see
#that tensorflow has some build in modules to auto do it
const = tf.constant(7) #A scalar constant
const

<tf.Tensor: shape=(), dtype=int32, numpy=7>

In [ ]:
#Check the number of dimensions using ndim
const.ndim

0

In [ ]:
#Creating vector
vector = tf.constant([10,10])
vector

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([10, 10], dtype=int32)>

In [ ]:
#Check vector dimensions
vector.ndim

1

In [ ]:
#Creating a matrix
matrix = tf.constant([[5,3],[3,5]])
matrix

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[5, 3],
       [3, 5]], dtype=int32)>

In [ ]:
matrix.ndim #matrix dimensions

2

In [ ]:
#New matrix
new_matrix = tf.constant([[5.,3.],[3.,5.]], dtype = tf.float16) #specify the data type
#By default it is int32, float16 is 16 precision, thus less memory
#If we get data type error ever, we can manipulate it using this
new_matrix

<tf.Tensor: shape=(2, 2), dtype=float16, numpy=
array([[5., 3.],
       [3., 5.]], dtype=float16)>

In [ ]:
#Dimension of the new matrix
new_matrix.ndim #these no. of dimensions are number of numbers inside a shape()
# 2,2 thus, 2 numbers of 2 dimensions, in short highest no. of index in shape() = no. of dimensions


2

In [ ]:
# Creating a tensor
tensor = tf.constant([[1,2,3],[1,2,3],[1,2,3]])
tensor

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[1, 2, 3],
       [1, 2, 3],
       [1, 2, 3]], dtype=int32)>

In [ ]:
#Number of dimensions
tensor.ndim

2

Thus,<br>
1) Scalar: a constant number<br>
2) Vector: a number and it's dimension<br>
3) Matrix: a 2D array of numbers<br>
4) Tensor: an n dimensional array of       numbers<br>
 * if n is 0 then it is a scalar
 * if n is 1 then it is a vector
 * if n is 2 then it is a matrix

In [ ]:
# New way to create tensors using 'tf.Variable'

In [ ]:
#Change able tensor vs unchanageable tensors
changeable = tf.Variable([5,3])
unchangeable = tf.constant([5,3])
changeable,unchangeable

(<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([5, 3], dtype=int32)>,
 <tf.Tensor: shape=(2,), dtype=int32, numpy=array([5, 3], dtype=int32)>)

In [ ]:
# The method that you could think of, but won't work
# changeable[0] = 1
# changeable

In [ ]:
# The correct way to do that use .assign()
changeable[0].assign(1)
changeable

<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([1, 3], dtype=int32)>

In [ ]:
# What happens if unchangable is tried to be manipulated
# unchangeable[0].assign(1)
# unchangeable

🔑**Note**: Rarely in practise you need to decide between tf.Variable and tf.constant, as TensorFlow does it for us. However, if not sure, use tf.constant, and change if needed later

# Creating Random Tensors

Random tensors are of random size with random numbers

In [ ]:
# Now that we know changeable and unchangeable
# Now, Create 2 Random tensors

random_one = tf.random.Generator.from_seed(42) #set seed for reproducibility
random_one = random_one.normal(shape=(3,2)) # chooses from a normal distribution
#with the shape of (3,2) meaning for an ease 3 rows, each with 2 elements

random_2 = tf.random.Generator.from_seed(42)
random_2 = random_2.normal(shape=(3,2))

#Equality check
random_one, random_2, random_one == random_2

(<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702],
        [ 0.07595026, -1.2573844 ],
        [-0.23193763, -1.8107855 ]], dtype=float32)>,
 <tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702],
        [ 0.07595026, -1.2573844 ],
        [-0.23193763, -1.8107855 ]], dtype=float32)>,
 <tf.Tensor: shape=(3, 2), dtype=bool, numpy=
 array([[ True,  True],
        [ True,  True],
        [ True,  True]])>)

They are pseudo random because we set a seed, with same seed as 42 in both, they are the same, changing the number in the seed will yield different result

#Shuffle the order of elements in a tensor

In [ ]:
# For example to learn multiple images at a same, it is useful
# Valueable for when you want to shuffle your data so the inherent order doesn't effect learning
not_shuffled = tf.constant([[10,3],[3,5],[1,2]])
# not_shuffled.ndim

# shuffle our not_shuffled
tf.random.shuffle(not_shuffled)


<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  3],
       [ 3,  5],
       [ 1,  2]], dtype=int32)>

In [ ]:
# Shuffle with seed
tf.random.set_seed(42) # Global seed
tf.random.shuffle(not_shuffled, seed = 42) # Operational level seed

# As seen before seed will produce pseudo random, and ordering change stays same for a particular seed value
# If no seed is choosen, it will generate random seed

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  3],
       [ 3,  5],
       [ 1,  2]], dtype=int32)>

#Other ways to create tensors

In [ ]:
# Similar to numpy
tf.ones([2,3]) # creates the tensor of shape (2,3) with all ones

tf.zeros([2,2]) # creates with shape (2,2) all the zeros, also 
# we can write tf.zeros(shape = (2,2)), it will yield same result
# Same with .ones

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[0., 0.],
       [0., 0.]], dtype=float32)>

# Turn NumPy arrays into tensors
The main difference between numpy and tensorflow tensors is that tensors can be run on a GPU for faster numerical computing. Otherwise, they are pretty much the same

In [ ]:
# Converting NumPy arrays to tensors
import numpy as np 
A = np.arange(1, 25, dtype = np.int32) # create a NumPy array between 1 and 25
A

# Norm is X = tf.constant(some_matrix) # capital for a matrix or tensor
#         y = tf.constant(vector) # non capital for a vector

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24], dtype=int32)

In [ ]:
# Converting that array A of numpy into a tensor
A = tf.constant(A)
A

<tf.Tensor: shape=(24,), dtype=int32, numpy=
array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24], dtype=int32)>

In [ ]:
# Giving shape to our tensor
A = tf.constant(A, shape = (2,3,4)) # 2 would be first bracket, 3 would be second, and 4 would be last
# since 24 elements 2*3*4 = 24
# Keep in mind, whatever shape you give, all of it's product should be equal to total no. of elements
# in this case 2*3*4 = total no. of elements (To accomodate all elements)
A

# Pay attention to brackets to understand better, there are 4 elements in the most inside brackets
# 3 vectors/blocks within 2nd brackets
# 2 nested vectors/blocks within the 1st bracket
# Thus, first dimension is 2, second is 2 and third is 4

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]]], dtype=int32)>

In [ ]:
# Let's try the shape of 6*2*2 = 24
A = tf.constant(A, shape = (6,2,2))
A

<tf.Tensor: shape=(6, 2, 2), dtype=int32, numpy=
array([[[ 1,  2],
        [ 3,  4]],

       [[ 5,  6],
        [ 7,  8]],

       [[ 9, 10],
        [11, 12]],

       [[13, 14],
        [15, 16]],

       [[17, 18],
        [19, 20]],

       [[21, 22],
        [23, 24]]], dtype=int32)>

In [ ]:
# Now, let's try 2*4*3 = 24
A = tf.constant(A, shape = (3,8))
A, A.ndim


(<tf.Tensor: shape=(3, 8), dtype=int32, numpy=
 array([[ 1,  2,  3,  4,  5,  6,  7,  8],
        [ 9, 10, 11, 12, 13, 14, 15, 16],
        [17, 18, 19, 20, 21, 22, 23, 24]], dtype=int32)>, 2)

# 2) Getting information from tensors
Tensors have following attributes, the main tensor attributes are: <br>
  * Shape
  * Rank
  * Axis or Dimension
  * Size

In [ ]:
# Creating a rank 4 tensor i.e. 4 dimensions
rank4_tensor = tf.zeros([2,3,4,5]) # can also use zeros(shape = [])
rank4_tensor

<tf.Tensor: shape=(2, 3, 4, 5), dtype=float32, numpy=
array([[[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]],


       [[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]]], dtype=float32)>

In [ ]:
# Watch out the brackets to deduce which is which, let us see [0] element of the above tensor
rank4_tensor[0]

<tf.Tensor: shape=(3, 4, 5), dtype=float32, numpy=
array([[[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]]], dtype=float32)>

In [ ]:
# We can even keep going inside to see if our deduction was correct or to access the inside elements
rank4_tensor[0][0]

<tf.Tensor: shape=(4, 5), dtype=float32, numpy=
array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]], dtype=float32)>

In [ ]:
# one more of such example like above
rank4_tensor[0][0][2] # this should give the last 2nd row of zeros

<tf.Tensor: shape=(5,), dtype=float32, numpy=array([0., 0., 0., 0., 0.], dtype=float32)>

In [ ]:
# To get all the attributes respectively
(rank4_tensor.shape, # would be the same as give
rank4_tensor.ndim, # would be 4 as there are 4 indices in shape(2,3,4,5) or 4 no.s to say
tf.size(rank4_tensor)) # would be total number of elements, 2*3*4*5 = 120, 120 elements

(TensorShape([2, 3, 4, 5]), 4, <tf.Tensor: shape=(), dtype=int32, numpy=120>)

In [ ]:
# Printing our various attributes well to just visualize better
print("Datatype of each element: ", rank4_tensor.dtype)
print("Shape of the tensor: ",rank4_tensor.shape)
print("Number of dimensions of axis: ", rank4_tensor.ndim)
print("Number of elements of 1st axis or dimension: ", rank4_tensor.shape[0]) # would be 2 as shape[0] is 2
print("Number of elements of last axis or dimension: ", rank4_tensor.shape[-1]) # would be 5
print("Size of the tensor: ", tf.size(rank4_tensor))
print("Size which only gives number of total elements in the tensor: ", tf.size(rank4_tensor).numpy())

Datatype of each element:  <dtype: 'float32'>
Shape of the tensor:  (2, 3, 4, 5)
Number of dimensions of axis:  4
Number of elements of 1st axis or dimension:  2
Number of elements of last axis or dimension:  5
Size of the tensor:  tf.Tensor(120, shape=(), dtype=int32)
Size which only gives number of total elements in the tensor:  120


# Indexing Tensors
Tensors can be indexed similar to python lists

In [ ]:
# A little exercise, let's get first 2 elements of our 1st dimension from our rank4_tensor
print((rank4_tensor[0] # since it should in the 1st dimension out of 2
 [0] # again since it should be in the first dimension out of 3
 [0] # again since it would be in the first dimension out of 4
 [0:2])) # the first 2 of the last dimension out of 5

print('\n')

 # Another way
print("Another way \n")
rank4_tensor[0,0,0, :2] # if we leave it blank instead of 0, then it will return the whole thing 0 means getting first index

tf.Tensor([0. 0.], shape=(2,), dtype=float32)


Another way 



<tf.Tensor: shape=(2,), dtype=float32, numpy=array([0., 0.], dtype=float32)>

In [ ]:
# Now lets get the 1st element from each dimension from each index except for the last one
(rank4_tensor[:1,:1,:1,:])

<tf.Tensor: shape=(1, 1, 1, 5), dtype=float32, numpy=array([[[[0., 0., 0., 0., 0.]]]], dtype=float32)>

In [ ]:
# Let's create a rank 2 tensor (2 dimensional)(rank = dimensions)
rank2_tensor = tf.constant([[1,2],[2,3]])
print("Shape: ", rank2_tensor.shape, "\nArray: ", rank2_tensor.numpy()) 
# .numpy will skip the extra info and give us only the array and dtype

print("\nWithout .numpy(): ", rank2_tensor)

Shape:  (2, 2) 
Array:  [[1 2]
 [2 3]]

Without .numpy():  tf.Tensor(
[[1 2]
 [2 3]], shape=(2, 2), dtype=int32)


In [ ]:
# Let's get the last 2 elements of each row of our 2d tensor
rank2_tensor[:,-1].numpy() # the first dimension of 2d tensor is row and 2nd dimension is column
# we needed last from each row, meaning last columns

# Simply we need -1 from each row, that is last element if each row, so row should be all
# and column should be -1

array([2, 3], dtype=int32)

In [ ]:
# How to add extra dimension to tensors
# Adding extra dimension to our 2d tensor for example
rank3_tensor = rank2_tensor[..., tf.newaxis] # dots represent all before, since we add at last
rank3_tensor

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[1],
        [2]],

       [[2],
        [3]]], dtype=int32)>

In [ ]:
# Alternative to above tf.newaxis method
tf.expand_dims(rank2_tensor, axis = -1) # -1 means expand the last axis
# change the axis=-1 to 0 or 1 wherevever you want to put an extra dimension
# it will add the extra dimension exactly where specified

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[1],
        [2]],

       [[2],
        [3]]], dtype=int32)>

# Manipulation of tensors or tensor operations
**Basic Operations** <br>
'+' , '-' , '*' , '/' <br>
 * The rules of matrices multiplication (also called dot product) must follow i.e.<br>
 1) The inner dimensions must match<br>
 2) The resulting matrix will have size of outer dimensions


In [ ]:
# Addition operator (+)
# add values to a tensor using '+'
tensor = tf.constant([[1,2],[2,3]])
print(tensor + 3) # will add 2 to all elements inside

print("Example to get a tensor with all 3s \n")
# for example to a tensor with all 3s
tensor = tf.zeros((2,2), dtype = tf.int16) #tf.dtype is used to change type,
# By default it was float 16 and remember (2,2) is equal to shape = (2,2)
print(tensor + 3)

tf.Tensor(
[[4 5]
 [5 6]], shape=(2, 2), dtype=int32)
Example to get a tensor with all 3s 

tf.Tensor(
[[3 3]
 [3 3]], shape=(2, 2), dtype=int16)


In [ ]:
# However the original tensor would be unchanged if we not assign it see below
tensor = tf.zeros((2,2), dtype = tf.int32)
print(tensor) #Gives all zeros, so to change it do the following
tensor = tensor + 3
print("\nThe changed tensor: ", tensor)

tf.Tensor(
[[0 0]
 [0 0]], shape=(2, 2), dtype=int32)

The changed tensor:  tf.Tensor(
[[3 3]
 [3 3]], shape=(2, 2), dtype=int32)


In [ ]:
# Mulitplication, same as +
tensor * 10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[30, 30],
       [30, 30]], dtype=int32)>

In [ ]:
# Substraction -
tensor - 2

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[1, 1],
       [1, 1]], dtype=int32)>

In [ ]:
# We can also use tensorflow built in functions for the same thing too
print(tf.multiply(tensor, 10)) # tf.math.multiply and tf.multiply are same thing
# The original tensor will remain unchanged
print("\n",tensor)
# To assign it just use tensor = tf.multiply(tensor, 10)

tf.Tensor(
[[30 30]
 [30 30]], shape=(2, 2), dtype=int32)

 tf.Tensor(
[[3 3]
 [3 3]], shape=(2, 2), dtype=int32)


To fully utlize gpu and better performance use tf.math.<"whatever you want">

**Matrix Multiplication** <br>
In machine learning, matrix mulitplication is one of the most common tensor operations.

In [ ]:
# Matrix multiplication in tensorflow
# multiplying a tensor/matrix with itself
print(tf.linalg.matmul(tensor, tensor)) # we can drop the middle linalg 
# for most of the tf.something1.something2 we can drop something1, it is okay

print("\nWithout .linalg and just tf.matmul: ", tf.matmul(tensor, tensor))

tf.Tensor(
[[18 18]
 [18 18]], shape=(2, 2), dtype=int32)

Without .linalg and just tf.matmul:  tf.Tensor(
[[18 18]
 [18 18]], shape=(2, 2), dtype=int32)


int16 would not work so make sure dtype of tensor is not int16

In [ ]:
# tensor * tensor would be element wise and would not yield matrix multiplication result 

# Little exercise
A = tf.constant([[1, 2, 5], [7, 2, 1], [3, 3, 3]])
B = tf.constant([[3, 5], [6, 7], [1, 8]])

AB = tf.matmul(A, B) # A*B matrix multiplication

print(AB.numpy()) # remember .numpy() to just get the result and no extra info

[[20 59]
 [34 57]
 [30 60]]


In [ ]:
# matrix multiplication using '@' operator in python
A @ B

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[20, 59],
       [34, 57],
       [30, 60]], dtype=int32)>

In [ ]:
# Remainder, we can shape using tf.reshape
tf.reshape(B, shape = (2,3))

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[3, 5, 6],
       [7, 1, 8]], dtype=int32)>

In [ ]:
# Check the size after reshape
tf.reshape(B, shape = (2,3)).shape

TensorShape([2, 3])

In [ ]:
# Can do the same with the transpose
tf.transpose(B), tf.reshape(B, shape = (2,3))
# Can see the difference, reshape just arranges to make up for the given dimensions if possible
# meaning B was 3*2 = 6, and after reshape 2*3=6 so, this is adjustable by reshape
# Transpose changes columns to rows and rows to column (Swaps axes) just shape changes

(<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
 array([[3, 6, 1],
        [5, 7, 8]], dtype=int32)>,
 <tf.Tensor: shape=(2, 3), dtype=int32, numpy=
 array([[3, 5, 6],
        [7, 1, 8]], dtype=int32)>)

**The dot product**<br>
Matrix multiplication is also called as dot product.
One can perform matrix multiplication using: <br>
 * tf.mathmul()
 * tf.tensordot()

In [ ]:
# lets perform dot product which will require the trsnapose or reshape, this time we use transpose
A = tf.constant([[1, 2, 5], [7, 2, 1], [3, 3, 3]])
B = tf.constant([[3, 5, 6], [7, 1, 8]])
print(A.shape, B.shape) # since inner elements are not same i.e. 3 for A and 2 for B

# To multiply we will need same inner dimensions
# By swaping 2 and 3 in B or transpose in other words, we can get same inner dimension of 3


(3, 3) (2, 3)


In [ ]:
# Matrix multiplication using tf.tensordot()
print(tf.transpose(B))
print(tf.tensordot(A, tf.transpose(B), axes = 1)) # read documentation to learn more, but axes is need to provided or it thorows error


tf.Tensor(
[[3 7]
 [5 1]
 [6 8]], shape=(3, 2), dtype=int32)
tf.Tensor(
[[43 49]
 [37 59]
 [42 48]], shape=(3, 2), dtype=int32)


Generally you transpose to satistfy matrix multiplication rules, rather than reshape<br>
Generally we use tf.mathmul()

In [ ]:
# We can change tensor datatype, as we did earlier, but this time explicitely
print(A.dtype) # will show int32 by feault
A = tf.cast(A, dtype = tf.float16) # This tf.cast(matrix,dtype) is the way to do it
print(A.dtype) 

# float16 or 16 precision in general is typically faster in calculations, need less space

<dtype: 'float16'>
<dtype: 'float16'>


# Aggregating tensors
Aggregating tensors = condensing them from multiple values down to a smaller amount of values

In [ ]:
# To get absolute values
# Not the best way, but should know
A = tf.constant([-1, -3]) # A tensor
tf.abs(A) # will turn all elements to absolute or mode |x|

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([1, 3], dtype=int32)>

Forms of aggregation:
 * Get minimum
 * Get maximum
 * Get mean 
 * Get sum
 * Get variance
 * Get std deviation <br>
 of the tensor

In [ ]:
# Lets create a tensor and use it for all the forms
A = tf.constant([[[1,2,3], [7,3,5]], [[9,6,13], [11,15,14]]])
A

<tf.Tensor: shape=(2, 2, 3), dtype=int32, numpy=
array([[[ 1,  2,  3],
        [ 7,  3,  5]],

       [[ 9,  6, 13],
        [11, 15, 14]]], dtype=int32)>

In [ ]:
# Getting the max across all the dimensions
# tf.math.reduce_max(A) or:
tf.reduce_max(A) # Both gives same results and are the same

<tf.Tensor: shape=(), dtype=int32, numpy=15>

In [ ]:
# Getting max along an axis or dimension
tf.reduce_max(A, axis = 1) # This should give max along the second dimension which is 2 in this case
# As we see this gives the max row/block from 2nd dimension

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[ 7,  3,  5],
       [11, 15, 14]], dtype=int32)>

In [ ]:
# Getting max from each row/block
tf.reduce_max(A, axis = 2) # This will give from the 3rd axis which is 3, can be seen from shape()
# The output should be 3,7,13,15 because those are the max in their respective rows
# Axis also starts from 0 because it is indexing inside shape() of a particular tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 3,  7],
       [13, 15]], dtype=int32)>

In [ ]:
# Same way getting minimum
print("Minimum of all: ", tf.reduce_min(A)) # Axis not specified means 0th axis or all elements
print("\nMinimum of 2nd dimension that is 2 here in shape(): ", tf.reduce_min(A, axis = 1))
print("\nMinimum of the last dimension in shape() which is 2: ", tf.reduce_min(A, axis = 2))
# The abve line will give from each row


Minimum of all:  tf.Tensor(1, shape=(), dtype=int32)

Minimum of 1st dimension that is 2 here in shape():  tf.Tensor(
[[ 1  2  3]
 [ 9  6 13]], shape=(2, 3), dtype=int32)

Minimum of the last dimension in shape() which is 2:  tf.Tensor(
[[ 1  3]
 [ 6 11]], shape=(2, 2), dtype=int32)


In [ ]:
# Similarly getting the mean
print("Mean of the entire tensor: ", tf.reduce_mean(A)) # Remember nothing specified means axis = 0
print("\nMean along the 2nd dimension/axis of shape(): ", tf.reduce_mean(A, axis = 1))
print("\nMean of the last dimension: ", tf.reduce_mean(A, axis = 2))

Mean of the entire tensor:  tf.Tensor(7, shape=(), dtype=int32)

Mean along the 2nd dimension/axis of shape():  tf.Tensor(
[[ 4  2  4]
 [10 10 13]], shape=(2, 3), dtype=int32)

Mean of the last dimension:  tf.Tensor(
[[ 2  5]
 [ 9 13]], shape=(2, 2), dtype=int32)


In [ ]:
# Similarly for the sum
print("Sum of the entire tensor: ", tf.reduce_sum(A))
print("\nSum along the second dimension/axis: ", tf.reduce_sum(A, axis = 1))
print("\nSum along the last dimension/axis: ", tf.reduce_sum(A, axis = 2))

Sum of the entire tensor:  tf.Tensor(89, shape=(), dtype=int32)

Sum along the second dimension/axis:  tf.Tensor(
[[ 8  5  8]
 [20 21 27]], shape=(2, 3), dtype=int32)

Sum along the last dimension/axis:  tf.Tensor(
[[ 6 15]
 [28 40]], shape=(2, 2), dtype=int32)


If did not understand about the 2nd dimesion sums or averages, this is how it happens: <br>
* For and sum too it is index to index of respective row, meaning:<br> 
[1,2,3],[7,3,5] -> this is our first element in 2nd dimension, now: <br>
for the sum: 1+7=8, 2+3=5, 3+5=8, hence [8 5 8] in the sum
* For the mean: <br>
[1,2,3],[7,3,5] -> this is our first element in 2nd dimension, now: <br>
1+7/2=4, 2+3/2=2, 3+5/2=4, hence [4 2 4] in the mean
<br>
<br>
Hence, 1 with 7, 2 with 3, and 3 with 5 since they same index in corresponding row

In [ ]:
# Another way to create random tensor, but this time with numbers and no probability distribution
random = tf.constant(np.random.randint(0, 100, size = 50)) # Creates a tensor with 50 elements between 0 and 100
random

<tf.Tensor: shape=(50,), dtype=int64, numpy=
array([30, 53, 84, 26, 77, 65, 25, 62, 53,  1, 11, 40, 41, 11, 49, 78, 82,
       16,  7, 88, 33, 17,  9, 48, 28, 16, 18, 82, 78, 76, 48, 57, 16,  7,
       43, 28, 86, 78, 70, 36, 45, 76, 77, 35, 83, 30, 12, 87, 88, 92])>

In [ ]:
# Lets apply all the aggrigator forms to this tensor
# The shape is 50 0 as we can see, since 50 elements in 1 row or 1 pair of brackets [ ]
print("The maximum in the random tensor: ", tf.reduce_max(random).numpy())
print("\nThe minimum in the random tensor: ", tf.reduce_min(random).numpy())
print("\nThe mean of the random tensor: ", tf.reduce_mean(random).numpy())
print("\nThe sum of the random tensor: ", tf.reduce_sum(random).numpy()) # .numpy() to see clear/only numbers
# We do not use axis argument because there is only 1 axis, can verify by using .shape()

The maximum in the random tensor:  92

The minimum in the random tensor:  1

The mean of the random tensor:  47

The sum of the random tensor:  2398


In [ ]:
# Now the remaining 2, i.e. variance and std deviation
# The syntax is the same, lets find them both for the both random and A

# We need to cast to float to get them, it would not work with int32
A = tf.cast(A, dtype = tf.float16)
print("Variance of A tensor along all the dimensions/axis", tf.math.reduce_variance(A).numpy())
print("\nStd deviation of A along all the dimensions/axis", tf.math.reduce_std(A).numpy())
print("\nVariance of A tensor along the 2nd dimensions/axis", tf.math.reduce_variance(A, axis = 1).numpy())
print("\nStd deviation of A along all the dimensions/axis", tf.math.reduce_std(A, axis = 1).numpy())

random = tf.cast(random, dtype = tf.float16)
print("\nVariance of random tensor along all the dimensions/axis", tf.math.reduce_variance(random).numpy())
print("Std deviation of random along all the dimensions/axis", tf.math.reduce_std(random).numpy())
#Note that random tensor only has 1 dimension



Variance of A tensor along all the dimensions/axis 22.08

Std deviation of A along all the dimensions/axis 4.7

Variance of A tensor along the 2nd dimensions/axis [[ 9.    0.25  1.  ]
 [ 1.   20.25  0.25]]

Std deviation of A along all the dimensions/axis [[3.  0.5 1. ]
 [1.  4.5 0.5]]

Variance of random tensor along all the dimensions/axis 784.0
Std deviation of random along all the dimensions/axis 28.0


🔑**Key take away from above variance and std. deviation** <br>
 * tf.math.reduce_variance() -> variance
 * tf.math.reduce_std() -> std. deviation
 * Variance and std. deviation only works with real or complex numbers meaning no int. Cast it to dytpe float as I did above, if it is int <br>
 **Note: ** The newer verison of tensor flow throws error for tf.reduce_variance without .math in between, and same for the std. deviation <br> If something like this happens with other build ins, just use the middle like .math in this case, refer the tensorflow documentation

In [ ]:
# To counter above problems where it does not work with ints
# We need to import tensorflow_probability
import tensorflow_probability as tfp #tfp is the norm
print(tfp.stats.variance(A))
print("\n")
print(tfp.stats.variance(tf.cast(random, dtype = tf.int32)).numpy()) # 784, same as with float16 we found using tf.math.reduce_variance()

# Lets now convert to int again and see
print("After making it int: " , tfp.stats.variance(tf.cast(A, dtype = tf.int32)))
# Lets cast our A tensor back to int and try

tf.Tensor(
[[16.    4.   25.  ]
 [ 4.   36.   20.25]], shape=(2, 3), dtype=float16)


784
After making it int:  tf.Tensor(
[[16  4 25]
 [ 4 36 20]], shape=(2, 3), dtype=int32)


float32 normally should be used, it is a std types for the tensors

# Positional Maximum and minimum index and value

In [ ]:
# Lets create a random tensor using uniform distribution
tf.random.set_seed(42) # Global seed
random = tf.random.uniform(shape=[50]) # creats 50 element random tensor from uniform distribution
random
# With random.unfiorm or .normal or any distribution, inside ( ) is the shape()

<tf.Tensor: shape=(50,), dtype=float32, numpy=
array([0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
       0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
       0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
       0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
       0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ,
       0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
       0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
       0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
       0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
       0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043],
      dtype=float32)>

In [ ]:
# To get the index/dimensional position where maximum number resides
print(tf.argmax(random)) # argmax(tensor) gives index of the max value
print("To see more clearly: ", tf.argmax(random).numpy())

tf.Tensor(42, shape=(), dtype=int64)
To see more clearly:  42


In [ ]:
# To get the value
# Just simple indexing as we have seen above
random[tf.argmax(random)]

<tf.Tensor: shape=(), dtype=float32, numpy=0.9671384>

In [ ]:
# The discussed above method tf.reduce_max() also works the same, and is easy
tf.reduce_max(random) # Answer is the same as above, thus, both works

<tf.Tensor: shape=(), dtype=float32, numpy=0.9671384>

In [ ]:
# For the minimum
print(random[tf.argmin(random)].numpy())
print(tf.reduce_min(random).numpy())
print((random[tf.argmin(random)] == tf.reduce_min(random)).numpy())

0.009463668
0.009463668
True


# Squeezing a tensor (removing all single dimensions/axis)

In [ ]:
# Lets create a random tensor with multilple single axis in shape()
tf.random.set_seed(42)
random = tf.random.uniform(shape = [1,1,50]) # tf.constant( tf.random.uniform(shape = [50], shape = (1,1,50)) will give same result
random


<tf.Tensor: shape=(1, 1, 50), dtype=float32, numpy=
array([[[0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
         0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
         0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
         0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
         0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ,
         0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
         0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
         0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
         0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
         0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043]]],
      dtype=float32)>

In [ ]:
# Lets squeeze and remove extra 1 dimensions to just extract the essence
tf.squeeze(random)
# you can also assign this squeezed to a new tensor, but I do not feel the need to use more space
# if the original is important as well, then might consider assignimg to new tensor

<tf.Tensor: shape=(50,), dtype=float32, numpy=
array([0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
       0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
       0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
       0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
       0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ,
       0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
       0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
       0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
       0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
       0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043],
      dtype=float32)>

# One-hot encoding tensors
* It is a type of a Numerical encoding <br>
* Input -> Numerical encoding (A way to represent data in numbers) <br>
📑 machinelearningmastery  -> a great website for machine learning

In [ ]:
# Creating a list of indices
indices = [0, 1, 2]  # These numbers can represent anything red, blue, green for example
# Remember the list should only have numbers, what it represents is arbitrary

# One hot encode our indices named list
tf.one_hot(indices, depth = 3) # depth is just the dimension that our tensor would get
# depth is just the number of elements in our list, according to which our tensor dimension is decided
# 3*3 in this case

<tf.Tensor: shape=(3, 3), dtype=float32, numpy=
array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]], dtype=float32)>

⚓**Note:** see how it puts 1s at positions provided in the list indices, if you change numbers lets say [1,2,0] in indices list then, the 1s in the encoding will appear as: <br>
0 1 0<br>
0 0 1<br>
1 0 0<br>

In [ ]:
# Instead of 1s and 0s we can give it custom representation for example
tf.one_hot(indices, depth = 3, on_value = "yes", off_value = "no")
# We will rarely use it, using 1s and 0s is the norm

<tf.Tensor: shape=(3, 3), dtype=string, numpy=
array([[b'yes', b'no', b'no'],
       [b'no', b'yes', b'no'],
       [b'no', b'no', b'yes']], dtype=object)>

# Few more useful .math operations
* square
* log
* square root

In [ ]:
# A way to create tensor of conitunes numbers
A = tf.range(1,10) # creates a tensor of 1 to 10 numbers (notice it just has one dimension/axis)
A.numpy()

array([1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=int32)

In [ ]:
# To sqaure it
tf.square(A).numpy()

array([ 1,  4,  9, 16, 25, 36, 49, 64, 81], dtype=int32)

In [ ]:
# For sqaure root
tf.math.sqrt(tf.cast(A, dtype = tf.float32)).numpy() # non int dtype required
# also .math is required. Note how .math is required if a method is dtype specific

array([0.99999994, 1.4142134 , 1.7320508 , 1.9999999 , 2.236068  ,
       2.4494896 , 2.6457512 , 2.8284268 , 3.        ], dtype=float32)

In [ ]:
# To find log
tf.math.log(tf.cast(A, dtype = tf.float32)).numpy() # Log is also dtype specific
# int not allowed

array([0.       , 0.6931472, 1.0986123, 1.3862944, 1.609438 , 1.7917595,
       1.9459102, 2.0794415, 2.1972246], dtype=float32)

# 4) Tensors and Numpy
* TensorFlow interacts beautifully with NumPy arrays. <br>
* TensorFlow is build upon the tensors and NumPy is build upon the arrays. <br>
* Numpy is fundamental python library for computing

In [ ]:
# Create a tensor directly from a NumPy array
A = tf.constant(np.array([1, 2, 3]))
A 

<tf.Tensor: shape=(3,), dtype=int64, numpy=array([1, 2, 3])>

In [ ]:
# Turn back the tensor to a NumPy array
(np.array(A), # will convert it to a numpy array back again
 type(np.array(A))) # This will confirm it, not we did not assigned the change to A,
#so we have to use (np.array(A)) in the bracket

(array([1, 2, 3]), numpy.ndarray)

In [ ]:
# Another way
A.numpy(), type(A.numpy())
# We have seen .numpy() a lot to just display clear array elements to see
# However, if we assign it by for example A = A.numpy(), then it will change the tensor to numpy array

(array([1, 2, 3]), numpy.ndarray)

In [ ]:
# It is useful if some functionality does not work with tensor type but works with numpy
# here is a way to access a particular element in tensor using numpy to only get the number
A.numpy()[1] # should show 2 as at index 1 value is 2

2

In [ ]:
# Default types for each are different
A = tf.constant(np.array([1, 2, 3]))
B = tf.constant([1, 2, 3])

#Check the data type
A.dtype, B.dtype # A would be with 64 precison while B with 32 by default 

# Thus, converting from numpy array would have 64 precision

(tf.int64, tf.int32)

# 5. @tf.function
The tf.function API is used in TF2.0 to create graphs for eagerly executed code.

There are two ways you can use this.
1. As a decorator: Using @tf.function decorator before your code will create a graph for that piece of code.

2. As a callable function : In this method you can simply tf.function-ise an existing function to create a graph for that function.

# 6. Finding access to GPUs

In [3]:
# Lets see the physical devices the run identifies/uses using run timr
tf.config.list_physical_devices()

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]

In [4]:
# Can specific by using
tf.config.list_physical_devices("GPU")

[]

In [1]:
# Google colab provides free to use GPU or TPU, by default it would be just CPU
# TPU requires set up, so lets only focus on gpu for now
# Go to rumtime -> change runtime type, and choose GPU
import tensorflow as tf
tf.config.list_physical_devices()

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'),
 PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

In [2]:
# Alright, we have a cpu and a gpu
!nvidia-smi # This will show what type of gpu you are using

Wed Jun 15 12:11:05 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   46C    P8     9W /  70W |      3MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

🔑**Note:** If you have access to a CUDA-enabled GPU, TensorFlow will automatically use it when possible. We can see above it is a CUDA powered GPU, CUDA version: 11.2 (by the time I am writing this notebook)